exporter/otlphttpexporter/otlp.go

```diff
@@ -228,3 +232,81 @@ func readResponse(resp *http.Response) *status.Status {
 
 	return respStatus
 }
+
+func handlePartialSuccessResponse(resp *http.Response, partialSuccessHandler partialSuccessHandler) error {
+	if resp.ContentLength == 0 {
+		return nil
+	}
+
+	maxRead := resp.ContentLength
+	needsResize := false
+	if maxRead == -1 || maxRead > maxHTTPResponseReadBytes {
+		maxRead = maxHTTPResponseReadBytes
+		needsResize = true
+	}
+	protoBytes := make([]byte, maxRead)
+	n, err := io.ReadFull(resp.Body, protoBytes)
+
+	// No bytes read and an EOF error indicates there is no body to read.
+	if n == 0 && (err == nil || errors.Is(err, io.EOF)) {
+		return nil
+	}
+
+	// io.ReadFull will return io.ErrorUnexpectedEOF if the Content-Length header
+	// wasn't set, since we will try to read past the length of the body. If this
+	// is the case, the body will still have the full message in it, so we want to
+	// ignore the error and parse the message.
+	if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return err
+	}
+
+	// The pdata unmarshaling methods check for the length of the slice
+	// when unmarshaling it, so we have to trim down the length to the
+	// actual size of the data.
+	if needsResize {
```

Comment:
The goal here was to use the Content Length header if it was present, otherwise try to adapt to an invalid/missing header. If `resp.ContentLength != n`, then the proto unmarshaling will fail. Since we're limiting the body to 64 KiB, I'm fine just allocating the byte slice and resizing it, which would also remove the check for the Content Length header.